aiosmtpd/smtp.py

```diff
@@ -108,6 +145,35 @@ def __init__(self, handler,
         self.envelope = None
         self.transport = None
         self._handler_coroutine = None
+        if not auth_require_tls and auth_required:
+            warn("Requiring AUTH while not requiring TLS "
+                 "can lead to security vulnerabilities!")
+        self._auth_require_tls = auth_require_tls
+        if auth_callback is None:
+            self._auth_callback = login_always_fail
+        else:
+            self._auth_callback = auth_callback
+        self._auth_required = auth_required
+        self.authenticated = False
+        # Get hooks & methods to significantly speedup getattr's
+        self._auth_methods: Dict[str, _AuthMechAttr] = {
+            m.replace("auth_", ""): _AuthMechAttr(getattr(h, m), h is self)
+            for h in (self, handler)
+            for m in dir(h)
+            if m.startswith("auth_")
+        }
+        for m in (auth_exclude_mechanism or []):
+            self._auth_methods.pop(m, None)
+        self._handle_hooks: Dict[str, Callable] = {
+            m.replace("handle_", ""): getattr(handler, m)
+            for m in dir(handler)
+            if m.startswith("handle_")
+        }
+        self._smtp_methods: Dict[str, Any] = {
+            m.replace("smtp_", ""): getattr(self, m)
+            for m in dir(self)
+            if m.startswith("smtp_")
+        }
```

Comment:
Embedded in this block of changes is my attempt to make the code more performant: Instead of keep invoking O(n) `getattr` every time, let's just cache all those `getattr` results in an O(1) dict.